src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3835,24 +3838,35 @@ def _subroutine_handler(self, node, parent):
             sub_spec = _first_type_match(node.content,
                                          Fortran2003.Specification_Part)
             decl_list = sub_spec.content
-            # TODO this if test can be removed once fparser/#211 is fixed
-            # such that routine arguments are always contained in a
-            # Dummy_Arg_List, even if there's only one of them.
-            if isinstance(node, (Fortran2003.Subroutine_Subprogram,
-                                 Fortran2003.Function_Subprogram)) and \
-               isinstance(node.children[0].children[2],
-                          Fortran2003.Dummy_Arg_List):
-                arg_list = node.children[0].children[2].children
-            else:
-                # Routine has no arguments
-                arg_list = []
         except ValueError:
             # Subroutine has no Specification_Part so has no
-            # declarations. Continue with empty lists.
+            # declarations. Continue with empty list.
             decl_list = []
+
+        # TODO this if test can be removed once fparser/#211 is fixed
+        # such that routine arguments are always contained in a
+        # Dummy_Arg_List, even if there's only one of them.
+        if isinstance(node, (Fortran2003.Subroutine_Subprogram,
+                             Fortran2003.Function_Subprogram)) and \
+           isinstance(node.children[0].children[2],
+                      Fortran2003.Dummy_Arg_List):
+            arg_list = node.children[0].children[2].children
+        else:
+            # Routine has no arguments
             arg_list = []
-        finally:
-            self.process_declarations(routine, decl_list, arg_list)
+
+        if arg_list and not decl_list:
```

Comment:
This doesn't but it will be caught subsequently when the PSyIR is constructed. I'll add a test.

Reply:
OK, so I think my question is therefore why is this particular case not get caught when the PSyIR is constructed and if it is not, why is it not fixed there rather than being fixed here?

Reply:
Oh, I see. I think I was catching it early to give a more precise error message. However, I've just removed this now and slightly refined the error message that is raised.